src/openms/source/ANALYSIS/QUANTITATION/AbsoluteQuantitation.cpp

```diff
@@ -0,0 +1,438 @@
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Douglas McCloskey $
+// $Authors: Douglas McCloskey $
+// --------------------------------------------------------------------------
+
+#include <OpenMS/ANALYSIS/QUANTITATION/AbsoluteQuantitation.h>
+
+//Kernal classes
+#include <OpenMS/KERNEL/StandardTypes.h>
+#include <OpenMS/KERNEL/MSExperiment.h>
+#include <OpenMS/KERNEL/MSChromatogram.h>
+#include <OpenMS/KERNEL/FeatureMap.h>
+#include <OpenMS/KERNEL/MRMTransitionGroup.h>
+#include <OpenMS/KERNEL/MRMFeature.h>
+
+//Analysis classes
+#include <OpenMS/ANALYSIS/MAPMATCHING/TransformationModel.h>
+#include <OpenMS/ANALYSIS/TARGETED/TargetedExperiment.h>
+
+//Quantitation classes
+#include <OpenMS/ANALYSIS/QUANTITATION/AbsoluteQuantitationStandards.h>
+#include <OpenMS/ANALYSIS/QUANTITATION/AbsoluteQuantitationMethod.h>
+
+//Math classes
+#include <OpenMS/MATH/STATISTICS/StatisticFunctions.h>
+
+//Standard library
+#include <cstddef> // for size_t & ptrdiff_t
+#include <vector>
+#include <math.h>
+#include <algorithm>
+
+namespace OpenMS
+{
+  
+  AbsoluteQuantitation::AbsoluteQuantitation() :
+  DefaultParamHandler("AbsoluteQuantitation")
+  {
+    //todo:  see MRMTransitionGroupPicker.cpp
+  }
+  
+  AbsoluteQuantitation::~AbsoluteQuantitation()
+  {
+  }
+
+  void AbsoluteQuantitation::setQuantMethods(std::vector<AbsoluteQuantitationMethod>& quant_methods)
+  {
+    quant_methods_.clear();
+    String component_name,IS_component_name,feature_name;
+    for (size_t i = 0; i < quant_methods.size(); i++)
+    {
+      quant_methods[i].getComponentISFeatureNames(component_name,IS_component_name,feature_name);
+      quant_methods_[component_name] = quant_methods[i];
+    }
+  }
+
+  double AbsoluteQuantitation::calculateRatio(const Feature & component_1, const Feature & component_2, const String & feature_name)
+  {
+    double ratio = 0.0;
+    if (component_1.metaValueExists(feature_name) && component_2.metaValueExists(feature_name))
+    {
+      double feature_1 = component_1.getMetaValue(feature_name);
+      double feature_2 = component_2.getMetaValue(feature_name);
+      // std::cout <<  "ratio = " << ratio << "." << std::endl;
+      ratio = feature_1/feature_2;
+    } 
+    else if (component_1.metaValueExists(feature_name))
+    {
+      LOG_INFO << "Warning: no IS found for component " << component_1.getMetaValue("native_id") << ".";
```

Comment:
did you mean `"Warning: no IS found for component " << component_2.getMetaValue("native_id")`

Reply:
No, I want to alert the user that the IS for component_1 is missing.